File: src/deploy.clj
(ns deploy
  (:require file const path home index git
            [main]))

(defn commit-folder-to [v]
  (let [repo (:repo v)
        deployment-dir (:deployment-dir v)
        branch (:branch v)
        current-branch (git/current-branch repo)
        tmp-dir "/tmp/jake-site-deploy"]

    (println "saving current changes")
    (git/add-all repo)
    (git/stash repo)

    (println "copying deployment to tmp dir")
    (git/checkout repo branch)
    (file/move deployment-dir tmp-dir repo)
    (git/status repo)

    (println "removing all untracked files")
    (git/remove-untracked repo)
    (git/status repo)

    (println "moving tmp dir contents to root")

    (file/copy-dir tmp-dir deployment-dir repo)
    (file/copy-dir (str deployment-dir "/*") "." repo)
    (git/status repo)

    (println "pushing build")
    (println "we are on branch " (git/current-branch repo))
    (git/add-all repo)
    (git/commit repo)
    (git/push repo)
    (git/status repo)

    (println "restoring working branch")
    (git/checkout repo current-branch)
    (git/status repo)

    (print "removing untracked")
    (git/remove-untracked repo)
    (git/status repo)

    (print "removing deployment dir")
    (file/remove-dir deployment-dir repo)
    (print "moving tmp dir to deployment dir")
    (file/move tmp-dir deployment-dir repo)

    (git/stash-pop repo)))

(defn -main [_]
  (let [current-repo "/home/jake/site"
        deployment-branch const/deployment-branch
        target-dir const/target-dir]
    (main/-main nil)
    (commit-folder-to {:repo current-repo
                       :branch deployment-branch
                       :deployment-dir target-dir})))

File: src/deploy.ts
// this file manages deployment to an external service
// right now its just github pages lol

import type { Repo } from "./utils/git";
import { Path } from "./utils/path";

function commitFolderToBranch({
  repo,
  folderToCommit,
  targetBranch,
}: {
  repo: Repo;
  folderToCommit: string;
  targetBranch: string;
}) {
  const tmpDir = "/tmp/jake-site-deploy";

  const currentBranch = repo.currentBranch();

  console.log("current branch", currentBranch);

  console.log("saving current changes");
  repo.addAll();
  repo.stash();

  console.log("copying deployment to tmp dir");

  // TODO: code should make sure 'production' branch is fetched.
  // it wasn't fetched! or the variable was not defined?
  repo.path.move(folderToCommit, tmpDir, { force: true });
  repo.checkout(targetBranch);
  repo.status();

  console.log("removing all untracked files");
  repo.removeUntracked();
  repo.status();

  console.log("moving tmp dir contents to root");

  Path.create(tmpDir).move(
    `${tmpDir}/*`,
    `${Path.create(folderToCommit).parent.toString()}/`,
    { force: true }
  );
  repo.status();

  console.log("pushing build");
  console.log("we are on branch ", repo.currentBranch());
  repo.addAll();
  repo.status();
  repo.commit();
  repo.push();
  repo.status();

  console.log("restoring working branch");
  repo.checkout(currentBranch);
  repo.status();

  console.log("removing untracked");
  repo.removeUntracked();
  repo.status();

  console.log("removing deployment dir");
  // TODO: file.removeDir(folderToCommit, repo);
  console.log("moving tmp dir to deployment dir");
  // TODO: file.move(tmpDir, folderToCommit, repo);

  repo.stashPop();
}

async function deploy({
  currentRepo,
  deploymentBranch,
  targetDir,
}: {
  currentRepo: Repo;
  deploymentBranch: string;
  targetDir: string;
}) {
  commitFolderToBranch({
    repo: currentRepo,
    targetBranch: deploymentBranch,
    folderToCommit: targetDir,
  });
}

export { deploy };
